### ISSUE TITLE
RAM consumption problem (no GC on some platform / python version)

### ISSUE THREAD

**Issue by xtpeqii** [fail2ban/fail2ban#1234]
According to my experience with both v 0.9.3(database disabled) and v0.8.4 using the same jails, actions and filters, after parsing some 9000000 one-line log failed records and banning some 700 ips, 0.8.4 consumes 3.7G RAM while 0.9.3 consumes 10G RAM. 
This is a little bit weird. Hope someone can help.

**Comment by sebres** [comment]
10G?!
Such leaks can occur in python, if we have some cross references in objects (that are unknown to me).
Then only one helps - execute `gc.collect()` occasionally (I think it should be done in each thread).

But I had never seen such leaks. 
Which python did you have? Which distribution (*nux version)?

**Comment by sebres** [comment]
Just looked at my biggest servers: f2b [ban-time-incr] >144 days up time, 10-15 jails, >10M log-lines processed, >100K failures, >10K bans:
```
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
# 1st
root      2195  0.2  0.1 232376  9616 ?        Sl   Jun06   3:34 /usr/bin/python /usr/local/bin/fail2ban-server -s /var/run/fail2ban/fail2ban.sock -p /var/run/fail2ban/fail2ban.pid -b
# 2nd
root      3006  0.1  0.1 129996  8172 ?        Sl   Jun03  15:33 /usr/bin/python /usr/local/bin/fail2ban-server -s /var/run/fail2ban/fail2ban.sock -p /var/run/fail2ban/fail2ban.pid -b
```
Can you provide me with output of folowing command in your system?
```bash
ps -eLf | grep 'fail2ban\|UID\s*PID'
```

**Comment by sebres** [comment]
Or better result of this:
```bash
 ps -eLo uid,pid,ppid,lwp,nlwp,vsz,osz,rss,ruser,pcpu,stime,etime,args | grep 'fail2ban\|UID\s*PID'
```

**Comment by sebres** [comment]
currently fast and small tests in the direction of mem-leak show: 
- `python 2.7`, `python 3.4` - no leak (I see nothing at all)
- `pypy3` - I see a small "leak", but I don't know that will be continued (memory manager overhead) and does not stop growing (possible it will be satiated at some point).

pypy was always very strange, what goes on object exemption or regarding to execution timing of the object destructor (`__del__`)
With other words: I have almost never needed to execute `gc.collect()` in python, but sometimes needed it in pypy, without cross references also.

**Comment by xtpeqii** [comment]
TO sebres:
thanks for your prompt replies. 
First I will answer your questjions:
Which python did you have?
A: python V.2.7.3
Which distribution (*nux version)?
A: CentOS release 5.5 (Final)

**Comment by xtpeqii** [comment]
Some extra info:
Before I start the fail2ban service ,I add 'ulimit -s 256' to the system service script /etc/init.d/fail2ban and put an echo at the end of the script to double check the statcksize is correctly set to 256KB.

Our servers are recieving billions of requests a day per server, and in just one day(the size of the logs per server is 3-7G), there are about 100 million matches per server, so I don't know if it's the way it is.

**Comment by xtpeqii** [comment]
After rerunning fail2ban for about 9 hours, 0.9.3 consume 3.8g  RAM, the output of the command you gave me was:
 UID   PID  PPID   LWP NLWP    VSZ SZ   RSS RUSER    %CPU STIME     ELAPSED COMMAND
    0  1769     1  1769    7 4224604 - 3987668 root    0.0 Oct29    15:31:10 /usr/local/bin/python2.7 /usr/local/bin/fail2ban-server -s /var/run/fail2ban/fail2ban.sock -p /var/run/fail2ban/fail2ban.pid -x -b
    0  1769     1  1778    7 4224604 - 3987668 root    3.5 Oct29    15:31:10 /usr/local/bin/python2.7 /usr/local/bin/fail2ban-server -s /var/run/fail2ban/fail2ban.sock -p /var/run/fail2ban/fail2ban.pid -x -b
    0  1769     1  1779    7 4224604 - 3987668 root    0.0 Oct29    15:31:10 /usr/local/bin/python2.7 /usr/local/bin/fail2ban-server -s /var/run/fail2ban/fail2ban.sock -p /var/run/fail2ban/fail2ban.pid -x -b
    0  1769     1  1780    7 4224604 - 3987668 root    2.1 Oct29    15:31:10 /usr/local/bin/python2.7 /usr/local/bin/fail2ban-server -s /var/run/fail2ban/fail2ban.sock -p /var/run/fail2ban/fail2ban.pid -x -b
    0  1769     1  1781    7 4224604 - 3987668 root    0.0 Oct29    15:31:10 /usr/local/bin/python2.7 /usr/local/bin/fail2ban-server -s /var/run/fail2ban/fail2ban.sock -p /var/run/fail2ban/fail2ban.pid -x -b
    0  1769     1  1782    7 4224604 - 3987668 root    2.2 Oct29    15:31:10 /usr/local/bin/python2.7 /usr/local/bin/fail2ban-server -s /var/run/fail2ban/fail2ban.sock -p /var/run/fail2ban/fail2ban.pid -x -b
    0  1769     1  1783    7 4224604 - 3987668 root    0.0 Oct29    15:31:10 /usr/local/bin/python2.7 /usr/local/bin/fail2ban-server -s /var/run/fail2ban/fail2ban.sock -p /var/run/fail2ban/fail2ban.pid -x -b
    0 15217  5946 15217    1  61180  -   816 root      0.0 09:07       00:00 grep fail2ban\|UID\s*PID

**Comment by sebres** [comment]
Strange... I should research further.

Can you please provide me a list jails you used and status of all jails?
```bash
for j in $(fail2ban-client status | grep 'Jail list:' | sed -r 's/[^:]+:\s*(.*$)/\1/' | tr " ," " "); do 
  echo $j; fail2ban-client status $j | grep -v 'Banned IP list'; 
done
```

**Comment by xtpeqii** [comment]
BELOW IS THE RESULT：
###########################
jishi-index
Status for the jail: jishi-index
|- Filter
|  |- Currently failed:	10
|  |- Total failed:	567304
|  `- File list:	/usr/local/webserver/tengine-2.1.0/logs/access.log
`- Actions
   |- Currently banned:	24
   |- Total banned:	24
jishi-login
Status for the jail: jishi-login
|- Filter
|  |- Currently failed:	13
|  |- Total failed:	8820116
|  `- File list:	/usr/local/webserver/tengine-2.1.0/logs/access.log
`- Actions
   |- Currently banned:	161
   |- Total banned:	161
jishi-queryuser
Status for the jail: jishi-queryuser
|- Filter
|  |- Currently failed:	4
|  |- Total failed:	181481
|  `- File list:	/usr/local/webserver/tengine-2.1.0/logs/access.log
`- Actions
   |- Currently banned:	26
   |- Total banned:	26
################

**Comment by sebres** [comment]
8.820.116 failures!? can be a candidate... (ca 400 bytes per failure, but it should have fewer tickets)
Which `findtime` and `maxretry` has jail jishi-login?
Can provide a `failreex` this jail used?

**Comment by xtpeqii** [comment]
There are 3 jails in my conf
For jishi-login: the findtime of them is 3 seconds, and the maxretry: 25
For the other two: 3seconds and 50 maxretry
jishi-login regex:
failregex = /acc/login\.html.*\t<HOST>\t\S+\t\S+$

**Comment by xtpeqii** [comment]
and now the RAM consumption is horribly : 19G..
some statistics:
jishi-index
Status for the jail: jishi-index
|- Filter
|  |- Currently failed:	7
|  |- Total failed:	7725845
|  `- File list:	/usr/local/webserver/tengine-2.1.0/logs/access.log
`- Actions
   |- Currently banned:	234
   |- Total banned:	772
jishi-login
Status for the jail: jishi-login
|- Filter
|  |- Currently failed:	16
|  |- Total failed:	31945718
|  `- File list:	/usr/local/webserver/tengine-2.1.0/logs/access.log
`- Actions
   |- Currently banned:	529
   |- Total banned:	1484
jishi-queryuser
Status for the jail: jishi-queryuser
|- Filter
|  |- Currently failed:	6
|  |- Total failed:	6656948
|  `- File list:	/usr/local/webserver/tengine-2.1.0/logs/access.log
`- Actions
   |- Currently banned:	269
   |- Total banned:	848

**Comment by sebres** [comment]
Jail `jishi-login` has totally 31945718 failures and 1484 bans
31945718 / 1484 = 21526 (average) failures per ban. Either `findtime` is too short or `maxretry` is to big.
Of course it should not produce such memory leak, but nevertheless... JFYI.
I suspect that all the failures that go not into BanManager will be not released or because a garbage collect will be not executed become the leak.
It's worse, that I cannot reproduce it with easy test case, just send failures to FailManager.
But I am still there and searching.

**Comment by sebres** [comment]
I have made a simple test case
```python
class AddFailure(unittest.TestCase):
...
	def testMemLeak(self):
		import gc
		class TestRef(object):
			def __init__(self, s):
				self._refs = []
				self.s = s
			def add_ref(self, ref):
				self._refs.append(ref)
			def __del__(self):
				# for ref in self._refs:
				# 	print('  has ref %s' % ref.s)
				pass

		# without cross reference ...
		a = TestRef('a')
		b = TestRef('b')
		a.add_ref(b);
		a.add_ref(TestRef('c'));

		a = None
		b = None

		# python (but not pypy) ...
		self.assertFalse(gc.collect())

		# make cross reference ...
		a = TestRef('a')
		b = TestRef('b')
		c = TestRef('c')
		a.add_ref(b);
		b.add_ref(c);
		c.add_ref(a);
		a = None
		b = None
		c = None
		# both python and pypy:
		self.assertTrue(gc.collect())
	
	def testFailuresMemLeak(self):
		import gc
		gc.collect()
		timestamp = 1167606999.0
		i2 = i3 = i4 = 0
		for i in xrange(0, 100000):
			i4 += 1
			if i4 > 255:
				i4 = 0
				i3 += 1
				if i3 > 255:
					i3 = 0
					i2 += 1
					if i2 > 255:
						i2 = 0
			t = FailTicket('127.%d.%d.%d' % (i2, i3, i4), 1167605999.0, dict({'match': i}))
			self.__failManager.addFailure(t)
			if i % 100 == 0:
				self.__failManager.cleanup(timestamp)
			self.__failManager.cleanup(timestamp)
		self.assertEqual(self.__failManager.size(), 0)
		t = i = i2 = i3 = i4 = None
		self.assertFalse(gc.collect())
```
But unfortunately nothing ... nary object was lost.
```
/bin/fail2ban-testcases MemLeak
Fail2ban 0.9.3.dev test suite. Python 2.7.6 (default, Jun 22 2015, 17:58:13) [GCC 4.8.2]. Please wait...
..
----------------------------------------------------------------------
Ran 2 tests in 2.150s

OK
```

**Comment by sebres** [comment]
I think I would need all the info about jails, please provide an output of:
```
fail2ban-client -d | grep jishi
```

**Comment by xtpeqii** [comment]
['add', 'jishi-login', 'auto']
['set', 'jishi-login', 'usedns', 'no']
['set', 'jishi-login', 'addlogpath', '/usr/local/webserver/tengine-2.1.0/logs/access.log', 'head']
['set', 'jishi-login', 'maxretry', 25]
['set', 'jishi-login', 'addignoreip', '127.0.0.1/8']
['set', 'jishi-login', 'addignoreip', '192.168.95.1/8']
['set', 'jishi-login', 'addignoreip', '205.147.11.34']
['set', 'jishi-login', 'addignoreip', '205.147.11.36']
['set', 'jishi-login', 'addignoreip', '180.117.59.23']
['set', 'jishi-login', 'addignoreip', '180.107.73.177']
['set', 'jishi-login', 'addignoreip', '221.224.9.114']
['set', 'jishi-login', 'addignoreip', '222.92.100.230']
['set', 'jishi-login', 'addignoreip', '58.211.28.178']
['set', 'jishi-login', 'addignoreip', '58.211.28.182']
['set', 'jishi-login', 'addignoreip', '222.73.95.135']
['set', 'jishi-login', 'addignoreip', '180.153.139.98']
['set', 'jishi-login', 'addignoreip', '222.73.95.135']
['set', 'jishi-login', 'logencoding', 'auto']
['set', 'jishi-login', 'bantime', 86400]
['set', 'jishi-login', 'ignorecommand', '']
['set', 'jishi-login', 'findtime', 3]
['set', 'jishi-login', 'addfailregex', '/acc/login\\.html.*\\t<HOST>\\t\\S+\\t\\S+$']
['set', 'jishi-login', 'addaction', 'jishi-login']
['set', 'jishi-login', 'action', 'jishi-login', 'actionban', '[[ `grep <ip> /usr/local/webserver/tengine-2.1.0/conf/denyconf/jishi-login.conf|wc -l` -eq 0 ]] && echo "if (\\$http_cdn_src_ip ~* \'<ip>\' ){ return 403; }"  >> /usr/local/webserver/tengine-2.1.0/conf/denyconf/jishi-login.conf\n/usr/local/webserver/tengine-2.1.0/sbin/nginx -s reload']
['set', 'jishi-login', 'action', 'jishi-login', 'actionstop', '']
['set', 'jishi-login', 'action', 'jishi-login', 'actionstart', 'touch /usr/local/webserver/tengine-2.1.0/conf/denyconf/jishi-login.conf']
['set', 'jishi-login', 'action', 'jishi-login', 'actionunban', 'grep -v <ip> /usr/local/webserver/tengine-2.1.0/conf/denyconf/jishi-login.conf > /usr/local/webserver/tengine-2.1.0/conf/denyconf/jishi-login.conf.tmp\nmv -f /usr/local/webserver/tengine-2.1.0/conf/denyconf/jishi-login.conf.tmp /usr/local/webserver/tengine-2.1.0/conf/denyconf/jishi-login.conf\n/usr/local/webserver/tengine-2.1.0/sbin/nginx -s reload']
['set', 'jishi-login', 'action', 'jishi-login', 'actioncheck', '']
['add', 'jishi-queryuser', 'auto']
['set', 'jishi-queryuser', 'usedns', 'no']
['set', 'jishi-queryuser', 'addlogpath', '/usr/local/webserver/tengine-2.1.0/logs/access.log', 'head']
['set', 'jishi-queryuser', 'maxretry', 50]
['set', 'jishi-queryuser', 'addignoreip', '127.0.0.1/8']
['set', 'jishi-queryuser', 'addignoreip', '192.168.95.1/8']
['set', 'jishi-queryuser', 'addignoreip', '205.147.11.34']
['set', 'jishi-queryuser', 'addignoreip', '205.147.11.36']
['set', 'jishi-queryuser', 'addignoreip', '180.117.59.23']
['set', 'jishi-queryuser', 'addignoreip', '180.107.73.177']
['set', 'jishi-queryuser', 'addignoreip', '221.224.9.114']
['set', 'jishi-queryuser', 'addignoreip', '222.92.100.230']
['set', 'jishi-queryuser', 'addignoreip', '58.211.28.178']
['set', 'jishi-queryuser', 'addignoreip', '58.211.28.182']
['set', 'jishi-queryuser', 'addignoreip', '222.73.95.135']
['set', 'jishi-queryuser', 'addignoreip', '180.153.139.98']
['set', 'jishi-queryuser', 'addignoreip', '222.73.95.135']
['set', 'jishi-queryuser', 'logencoding', 'auto']
['set', 'jishi-queryuser', 'bantime', 86400]
['set', 'jishi-queryuser', 'ignorecommand', '']
['set', 'jishi-queryuser', 'findtime', 3]
['set', 'jishi-queryuser', 'addfailregex', '/9yin/queryUser\\.html.*\\t<HOST>\\t\\S+\\t\\S+$']
['set', 'jishi-queryuser', 'addaction', 'jishi-queryuser']
['set', 'jishi-queryuser', 'action', 'jishi-queryuser', 'actionban', '[[ `grep <ip> /usr/local/webserver/tengine-2.1.0/conf/denyconf/jishi-queryuser.conf|wc -l` -eq 0 ]] && echo "if (\\$http_cdn_src_ip ~* \'<ip>\' ){ return 403; }"  >> /usr/local/webserver/tengine-2.1.0/conf/denyconf/jishi-queryuser.conf\n/usr/local/webserver/tengine-2.1.0/sbin/nginx -s reload']
['set', 'jishi-queryuser', 'action', 'jishi-queryuser', 'actionstop', '']
['set', 'jishi-queryuser', 'action', 'jishi-queryuser', 'actionstart', 'touch /usr/local/webserver/tengine-2.1.0/conf/denyconf/jishi-queryuser.conf']
['set', 'jishi-queryuser', 'action', 'jishi-queryuser', 'actionunban', 'grep -v <ip> /usr/local/webserver/tengine-2.1.0/conf/denyconf/jishi-queryuser.conf > /usr/local/webserver/tengine-2.1.0/conf/denyconf/jishi-queryuser.conf.tmp\nmv -f /usr/local/webserver/tengine-2.1.0/conf/denyconf/jishi-queryuser.conf.tmp /usr/local/webserver/tengine-2.1.0/conf/denyconf/jishi-queryuser.conf\n/usr/local/webserver/tengine-2.1.0/sbin/nginx -s reload']
['set', 'jishi-queryuser', 'action', 'jishi-queryuser', 'actioncheck', '']
['add', 'jishi-index', 'auto']
['set', 'jishi-index', 'usedns', 'no']
['set', 'jishi-index', 'addlogpath', '/usr/local/webserver/tengine-2.1.0/logs/access.log', 'head']
['set', 'jishi-index', 'maxretry', 50]
['set', 'jishi-index', 'addignoreip', '127.0.0.1/8']
['set', 'jishi-index', 'addignoreip', '192.168.95.1/8']
['set', 'jishi-index', 'addignoreip', '205.147.11.34']
['set', 'jishi-index', 'addignoreip', '205.147.11.36']
['set', 'jishi-index', 'addignoreip', '180.117.59.23']
['set', 'jishi-index', 'addignoreip', '180.107.73.177']
['set', 'jishi-index', 'addignoreip', '221.224.9.114']
['set', 'jishi-index', 'addignoreip', '222.92.100.230']
['set', 'jishi-index', 'addignoreip', '58.211.28.178']
['set', 'jishi-index', 'addignoreip', '58.211.28.182']
['set', 'jishi-index', 'addignoreip', '222.73.95.135']
['set', 'jishi-index', 'addignoreip', '180.153.139.98']
['set', 'jishi-index', 'addignoreip', '222.73.95.135']
['set', 'jishi-index', 'logencoding', 'auto']
['set', 'jishi-index', 'bantime', 86400]
['set', 'jishi-index', 'ignorecommand', '']
['set', 'jishi-index', 'findtime', 3]
['set', 'jishi-index', 'addfailregex', '/9yin/index\\.html.*\\t<HOST>\\t\\S+\\t\\S+$']
['set', 'jishi-index', 'addaction', 'jishi-index']
['set', 'jishi-index', 'action', 'jishi-index', 'actionban', '[[ `grep <ip> /usr/local/webserver/tengine-2.1.0/conf/denyconf/jishi-index.conf|wc -l` -eq 0 ]] && echo "if (\\$http_cdn_src_ip ~* \'<ip>\' ){ return 403; }"  >> /usr/local/webserver/tengine-2.1.0/conf/denyconf/jishi-index.conf\n/usr/local/webserver/tengine-2.1.0/sbin/nginx -s reload']
['set', 'jishi-index', 'action', 'jishi-index', 'actionstop', '']
['set', 'jishi-index', 'action', 'jishi-index', 'actionstart', 'touch /usr/local/webserver/tengine-2.1.0/conf/denyconf/jishi-index.conf']
['set', 'jishi-index', 'action', 'jishi-index', 'actionunban', 'grep -v <ip> /usr/local/webserver/tengine-2.1.0/conf/denyconf/jishi-index.conf > /usr/local/webserver/tengine-2.1.0/conf/denyconf/jishi-index.conf.tmp\nmv -f /usr/local/webserver/tengine-2.1.0/conf/denyconf/jishi-index.conf.tmp /usr/local/webserver/tengine-2.1.0/conf/denyconf/jishi-index.conf\n/usr/local/webserver/tengine-2.1.0/sbin/nginx -s reload']
['set', 'jishi-index', 'action', 'jishi-index', 'actioncheck', '']
['start', 'jishi-login']
['start', 'jishi-queryuser']
['start', 'jishi-index']

**Comment by sebres** [comment]
Very strange. Indeed you have an custom banaction (denyconf/nginx-reload) but it should be not the problem. Especially because you have many failures and much fewer bans (that also can not produce such leaks, I can just not imagine that).
I am convinced that such extent leak indicates a something wrong by the handling with failures. But I not see that...

It will really help If you could repeat my test case above:
Just add testFailuresMemLeak def to class AddFailure inside a fail2ban/tests/failmanagertestcase.py for example like here - https://github.com/sebres/fail2ban/blob/_tmp/mem-leak/fail2ban/tests/failmanagertestcase.py#L87
And then start the test cases with
```
cd ./f2b-path
PYTHONPATH=. bin/fail2ban-testcases MemLeak
```
In line https://github.com/sebres/fail2ban/blob/_tmp/mem-leak/fail2ban/tests/failmanagertestcase.py#L93 you can also try to increase the repeat counter to xrange(0, 1000000) and track the memory usage (MEM%/VSZ/RSS) of the process running python with tests in separate session via top/htop.

**Comment by xtpeqii** [comment]
I add gc.collect()  to  failmanager.py ：
```
# emacs: -*- mode: python; py-indent-offset: 4; indent-tabs-mode: t -*-
# vi: set ft=python sts=4 ts=4 sw=4 noet :

# This file is part of Fail2Ban.
#
# Fail2Ban is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# Fail2Ban is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with Fail2Ban; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

# Author: Cyril Jaquier
# 
.....
......
	def addFailure(self, ticket):
                gc.collect()
....
```
Hope it helps...

**Comment by xtpeqii** [comment]
To sebres, can you give me any suggestions about where to add garbage collector code?
MUCH THANKS

**Comment by sebres** [comment]
My suggestion was the execution of the test case, I made.

But I would have added it in FailManager::cleanup, see https://github.com/fail2ban/fail2ban/blob/master/fail2ban/server/failmanager.py#L129
```diff
+ import gc
import logging
....
	def cleanup(self, time):
		try:
			self.__lock.acquire()
			tmp = self.__failList.copy()
			for item in tmp:
				if tmp[item].getLastTime() < time - self.__maxTime:
					self.__delFailure(item)
		finally:
			self.__lock.release()
+		gc.collect()
```

**Comment by xtpeqii** [comment]
OK, after adding some magic to failmanager.py as I posted above, I finally managed to make fail2ban not hungry for RAM(118MB) any more.  And now the output is:
```
jishi-index
Status for the jail: jishi-index
|- Filter
|  |- Currently failed:	51
|  |- Total failed:	500047
|  `- File list:	/usr/local/webserver/tengine-2.1.0/logs/access.log
`- Actions
   |- Currently banned:	88
   |- Total banned:	88
jishi-login
Status for the jail: jishi-login
|- Filter
|  |- Currently failed:	33
|  |- Total failed:	3680997
|  `- File list:	/usr/local/webserver/tengine-2.1.0/logs/access.log
`- Actions
   |- Currently banned:	1006
   |- Total banned:	1006
jishi-queryuser
Status for the jail: jishi-queryuser
|- Filter
|  |- Currently failed:	68
|  |- Total failed:	899875
|  `- File list:	/usr/local/webserver/tengine-2.1.0/logs/access.log
`- Actions
   |- Currently banned:	169
   |- Total banned:	169
```
Thank you very much ,sebres, for following my issue. 
Now the RAM consumption has dropped drastically by 99%. But the side effect is that the cpu load has rosen as fail2ban will run gc.collect() every time a failure occures. 
So next thing I  will try is to add gc.collect() to cleanup function as you suggested and keep an eye for the CPU load change.

**Comment by cfrancis2b** [comment]
I am experiencing the same problem, come to think that had something misconfigured but as discussed this happens since Version 0.9.3 was installed.

Please let us know if they find a solution.

Regards.

**Comment by sebres** [comment]
> since Version 0.9.3 was installed

Above which version have you updated?

I tend to believe that would be some version of python resp. platform issue (not directly fail2ban self). Just because I can't reproduce it for example on debian's / python 2.7.8 resp. python 3.3 and 3.4.
Or it is some constellation with fail2ban configuration...

Until now I don't understand why gc.collect will be expected by failures handling...
But I would like to understand, because the forced usage of gc.collect would be rather a workaround as solution.

**Comment by cfrancis2b** [comment]
In my case:
- Centos 6.0
- Python 2.6.6

For now the problem returning solvent to version 0.8.14.

**Comment by xtpeqii** [comment]
Now I'am here to present my fianl solution for my case.
Remember I posted that I would move `gc.collect()` to where Sebres suggested?
Unfortunately, it didn't help decrease the RAM consumption, it should work as both Sebres and I had imagined, but it didn't... 
Although putting `gc.collect()` to addFailure function in   failmanager.py did halt the crazy RAM usage, running `gc.collect()` added more load to CPU. 
So, finally, I came up with this,in  failmanager.py:
```
def addFailure(self, ticket):
***+                p=subprocess.Popen(['/bin/bash','-c','ps aux|grep "[f]ail2ban-server"|perl -ane \'print $F[5]\''],stdout=subprocess.PIPE,stderr=subprocess.STDOUT)
+                ramSize=p.communicate()
+                if int(ramSize[0])>1048576:
+                        gc.collect()***
                  try:
                        self.__lock.acquire()
```

**Comment by xtpeqii** [comment]
To cfrancis2b :
In my case, the 0.8.X version still has the RAM consumption problem, but it consumed less as the 1st floor of this post suggested.

**Comment by sebres** [comment]
as quick and dirty solution it can be alright, but we should make something better as it :)
for example much better and easy (and does not start any foreign processes) would be a simple latency after last gc.
Something like this (in example below it is 30 seconds):
```python
+  #todo: make latency (30) configurable:
+  if time > self.__next_gc:
+    gc.collect()
+    self.__next_gc = time + 30
```
Is also a quick "solution", but quite not so dirty :)

**Comment by xtpeqii** [comment]
TO sebres:
Unfortunately, the suggestion in your latest reply didn't work at all.....
Instead of adding the if block, I simply put gc.collect() there and it worked.
ridiculous as it sounds, it worked... I don't know why, neither can I understand why I found from a programming perspective, it just worked and made a defference compared with putting gc.collect() in a simple if block.

**Comment by sebres** [comment]
I said "something like this"... How familiar are you with python? 
For example, to use **exact** code above, you should first add ```self.__next_gc = -0x7fffffff``` in constructor (init) of the class.

**Comment by xtpeqii** [comment]
not so familiar...  just a beginner, actually I'm an SA trying to expand my capacity of programming

**Comment by sebres** [comment]
What we're doing here is a workaround: 
- the garbage collector servicing should occur full automatically. But something in your system prevents this (possibly a mainainer of your system, platform, fail2ban etc has completelly disabled GC resp. whose threshold is set very high (I strongly believe that);
- GC works in python cross-thread, so if we build a workaround it should act periodically once cross all threads;

I will make a better workaround for your issue.
In this time please provide an output of:
```bash
python -c 'import gc; print("GC - %s - %s" % (gc.isenabled(), gc.get_threshold()))'
```

**Comment by sebres** [comment]
Here you go - https://github.com/fail2ban/fail2ban/compare/master...sebres:_sb/bg-svc-gc-collect

To test it, you should replace 2 files in your fail2ban directory (mostly in something like `/usr/local/lib/python2.7/dist-packages/fail2ban`):
- fail2ban/helpers.py
- fail2ban/server/failmanager.py

If you are not familiar with git, you can download both files:
- https://raw.githubusercontent.com/sebres/fail2ban/f1db82df893bb16b87d8f3def66bb1448750c9d1/fail2ban/helpers.py
- https://raw.githubusercontent.com/sebres/fail2ban/f1db82df893bb16b87d8f3def66bb1448750c9d1/fail2ban/server/failmanager.py
- 3rd file in commit (fail2ban/tests/failmanagertestcase.py) is just a test case;

Or as zip:
https://github.com/sebres/fail2ban/archive/_sb/bg-svc-gc-collect.zip

**Comment by sebres** [comment]
Currently updated (and fixed links above also) - I've commited temporary test version without `gc.collect()`. (I'm sorry)

**Comment by xtpeqii** [comment]
thanks, sebres. After the end-of-year busy working days,  I will exert more effort in learning python.

**Comment by sebres** [comment]
You are welcome.
> I will exert more effort in learning python

It never hurts... :)

**Comment by sebres** [comment]
Still another situation (as missed GC) may occur with OOM result (see #1277), so everyone that has the same issue can test my branch https://github.com/sebres/fail2ban/tree/f2b-perfom-prepare-716, that should fix it (see https://github.com/fail2ban/fail2ban/issues/1277#issuecomment-166725561 also).

@xtpeqii can you possibly check it also, please?

**Comment by xrg** [comment]
I confirm high memory usage on Mageia 4 (similar to CentOS 7)  + systemd logs, asterisk, apache jails.

**Comment by yarikoptic** [comment]
Great to see all the activity here. Is this applicable to the mainline code as well?

On December 24, 2015 12:54:31 PM EST, Panos <notifications@github.com> wrote:
>I confirm high memory usage on Mageia 4 (similar to CentOS 7)  +
>systemd logs, asterisk, apache jails.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/fail2ban/fail2ban/issues/1234#issuecomment-167141561

-- 
Sent from a phone which beats iPhone.

**Comment by sebres** [comment]
@yarikoptic Conditionally, but question is with which effort... Let me see.

**Comment by xtpeqii** [comment]
@sebres  sorry, the attack has stopped, so there will not be so much data anymore unless another DDOS happens..

**Comment by sebres** [comment]
Closed, because it was merged in 0.10 (further in #1410)
